src/ZipFile.jl

```diff
@@ -352,7 +358,7 @@ function close(w::Writer)
 	_writele(w._io, @compat UInt32(cdpos))
 	_writele(w._io, @compat UInt16(0))
 
-	if !isa(w._io, IOBuffer)
+	if w._close_io
```

Comment:
There is no need to add an extra flag. Add a `flush` function and call `flush` from from inside `close`. If a user don't want to close the underlying IOBuffer, they'll just call flush instead of close.

Please send this as a separate pull request. It's easier to review that way.